### PR TITLE
refactor `devtools/td-layout-config` and `td-layout` crates

### PR DESCRIPTION
**Pull request by gaojiaqi7**
### Issue related to this PR

 - Fix: https://github.com/confidential-containers/td-shim/issues/461

### Refactor the `devtools/td-layout-config`

1. Merge `td-shim-layout-builder` under `td-shim-tools` into `devtools/td-layout-config`

Make some optimizations to support memory type which can help to determine the location of memory regions (high-end or low-end of low memory) and to generate E820 table at runtime with the output rust format configurations.

The memory layout configuration file will look like this:

```
{
    "memory_regions": [
        {
            "name": "Bootloader",
            "size": "0x800000",
            "type": "Memory"
        },
        {
            "name": "TdHob",
            "size": "0x20000",
            "type": "Memory"
        },
        {
            "name": "PayloadParam",
            "size": "0x40000",
            "type": "Memory"
        }
        ...
        
        {
            "name": "EventLog",
            "size": "0x100000",
            "type": "Nvs"
        },
        ...
    ]
}
```

Add the image layout configuration to the new tool which is not supported by `td-shim-layout-builder`.

2. Split the image configuration and memory configuration

Linux payload and excutable payload will use one same image configuration but different memory layout, so we can split image and memory configuration for now.

### Update the `td-layout` used by runtime of `td-shim`

1. Use one `RuntimeLayout` defination for different layouts

Previous implementation of `RuntimeLayout` defines fixed members which can only be used for one layout. In this PR, an array is used to hold memory regions with address, size and type informations and it can be initialized with layout configuration at runtime.

2. Move `get_dynamic_mem_slice()` to `Memory`

`Memory` struct holds the information of memory regions and it will be more natural and convenient to get the slice of these regions from `Memory` struct.

### CODE REVIEW

**Comment by gaojiaqi7** [comment]
#466 #461 #460

File: td-shim/src/bin/td-shim/memory.rs [hunk 1]
```diff
@@ -399,7 +399,7 @@ pub fn cpu_get_memory_space_size() -> u8 {
     size_of_mem_space
 }
 
-#[cfg(test)]
+// #[cfg(test)]
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/memory.rs:402]: why comment it?

**Comment by jyao1** [comment]
please split the patch to smaller ones. It is hard to review a refactoring patch with 66 files updated.

**Comment by gaojiaqi7** [comment]
Split this PR into PR https://github.com/confidential-containers/td-shim/pull/519, PR https://github.com/confidential-containers/td-shim/pull/516, PR https://github.com/confidential-containers/td-shim/pull/517 and PR PR https://github.com/confidential-containers/td-shim/pull/518

**Pull request closed without merging**